airbyte-integrations/connectors/source-square/source_square/components.py

```diff
@@ -0,0 +1,153 @@
+        expected: "expires_in": number of seconds
+        """
+        token, expires_at = super().refresh_access_token()
+        expires_in = pendulum.parse(expires_at) - pendulum.now()
+        return token, expires_in.in_seconds()
+
+
+@dataclass
+class AuthenticatorSquare(DeclarativeAuthenticator, JsonSchemaMixin):
+    config: Mapping[str, Any]
+    bearer: BearerAuthenticator
+    oauth: Oauth2AuthenticatorSquare
+
+    def __new__(cls, bearer, oauth, config, *args, **kwargs):
+        if config.get('api_key'):
+            return bearer
+        else:
+            return oauth
+
+
+@dataclass
+class SquareSlicer(StreamSlicer):
+    options: InitVar[Mapping[str, Any]]
+    request_cursor_field: str
+    cursor_field: str
+
+    START_DATETIME: ClassVar[str] = "1970-01-01T00:00:00+00:00"
+    DATETIME_FORMAT: ClassVar[str] = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+    def __post_init__(self, options: Mapping[str, Any]):
+        self._state = {}
+
+    def get_stream_state(self) -> StreamState:
+        return self._state
+
+    def get_request_params(
+            self,
+            *,
+            stream_state: Optional[StreamState] = None,
+            stream_slice: Optional[StreamSlice] = None,
+            next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+    def get_request_headers(self, *args, **kwargs) -> Mapping[str, Any]:
+        return {}
+
+    def get_request_body_data(self, *args, **kwargs) -> Optional[Union[Mapping, str]]:
+        return {}
+
+    def get_request_body_json(
+            self,
+            *,
+            stream_state: Optional[StreamState] = {},
+            stream_slice: Optional[StreamSlice] = None,
+            next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> Optional[Mapping]:
+
+        return {}
+
+    def _max_dt_str(self, *args: str) -> Optional[str]:
+        new_state_candidates = list(map(lambda x: pendulum.parse(x) if isinstance(x, str) else x, filter(None, args)))
+        if not new_state_candidates:
+            return
+        max_dt = max(new_state_candidates)
+        (dt, micro) = max_dt.strftime(self.DATETIME_FORMAT).split(".")
+        return "%s.%03dZ" % (dt, int(micro[:-1:]) / 1000)
+
+    def update_cursor(self, stream_slice: StreamSlice, last_record: Optional[Record] = None):
+        slice_state = stream_slice.get(self.cursor_field)
+        current_state = self._state.get(self.cursor_field)
+        last_cursor = last_record and last_record[self.cursor_field]
+        max_dt = self._max_dt_str(slice_state, current_state, last_cursor)
+        if not max_dt:
+            return
+        self._state[self.cursor_field] = max_dt
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState, *args, **kwargs) -> Iterable[StreamSlice]:
+        yield {self.request_cursor_field: stream_state.get(self.cursor_field, self.START_DATETIME)}
+
+
+@dataclass
+class SquareSubstreamSlicer(SquareSlicer):
```

Comment:
what missing features is this covering for? Would love to find a way to reuse existing components or make this generalizable enough that it can be reused

Reply:
This class is similar to `SubstreamSlicer`, and even can be used here, but it utilised 1:1 relation (parent:child). In my case I need to get all parent records (`locations` stream), get `id` from them and iterate in batch (10 ids in request)

Reply:
Why do we want to batch 10 the slices in batches of 10?

Reply:
![image](https://user-images.githubusercontent.com/36314070/203382493-18e12959-24e4-4053-b09b-55052996f3b7.png)
we want to batch 10 locations to optimise requests

Reply:
Could you create an issue advocating for adding this functionality into the lowcode CDK as described [here](https://airbytehq-team.slack.com/archives/C02U9R3AF37/p1670428588243999)? 

If I'm understanding correctly, the additional features needed from the substream slicer are: 
1. parsing a subfield of the parent record/slice
2. batching multiple parent streams in the output

same thing as mentioned above

Reply:
Done https://github.com/airbytehq/airbyte/issues/19760